src/cli.rs

```diff
@@ -1,22 +1,120 @@
 //use web3::{Web3, transports};
+use clap::{App, Arg};
 use reqwest::blocking as reqwest;
 use serde_json;
+use std::env;
+use std::ffi::OsString;
 use std::fs;
 use std::io::{self, Read, Write};
+use std::net::TcpListener;
+use std::net::TcpStream;
 use std::os::unix;
 use threadpool::ThreadPool;
 
-// TODO: things to configure:
-//  - infura project id (not just env var?)
-//  - rpc endpoint port
-//  - rpc endpoint type (tcp, ws, ipc)
-//  - max concurrent requests (ie~ threadpool size)
+#[derive(Debug, PartialEq)]
+pub struct TrinConfig {
+    pub protocol: String,
+    pub infura_project_id: String,
+    pub endpoint: u32,
+    pub pool_size: u32,
+}
+
+impl TrinConfig {
+    pub fn new() -> Self {
+        Self::new_from(std::env::args_os().into_iter()).unwrap_or_else(|e| e.exit())
+    }
+
+    fn new_from<I, T>(args: I) -> Result<Self, clap::Error>
+    where
+        I: Iterator<Item = T>,
+        T: Into<OsString> + Clone,
+    {
+        let matches = App::new("trin")
+            .version("0.0.1")
+            .author("carver")
+            .about("super lightweight eth portal")
+            .arg(
+                Arg::with_name("protocol")
+                    .short("p")
+                    .long("protocol")
+                    .help("select transport protocol")
+                    .takes_value(true)
+                    .default_value("http"),
+            )
+            .arg(
+                Arg::with_name("endpoint")
+                    .short("e")
+                    .long("endpoint")
```

Comment:
I like `"http_port"` - just `"hp"` won't be possible as the short flag. [short](https://docs.rs/clap/2.33.3/clap/struct.Arg.html#method.short) automatically strips whatever the argument is down to just the first level. This made it kinda hard to come up with neat intuitive short flags. Though it seems to me like supporting a confusing short flag is better than no short flags? Also check out that the short flag for `"pool_size"` is currently `"s"` (since `"protocol"` gets `"p"`). But I'm happy to switch to whatever naming convention you feel makes sense.